### ISSUE TITLE
Can not write stream into jdbc sink unless key is set

### ISSUE THREAD

**Issue by holgerbrandl** [confluentinc/ksql#3487]
Extending the Quickstart example from https://docs.confluent.io/3.1.1/connect/connect-jdbc/docs/sink_connector.html#quickstart, the user may want to cretate a ksql stream and write it into jdbc.

However, this seems to fail using the confluent local distribution v5.3.1 when using the following protocol:

```
# create ksql stream from orders
ksql <<EOF
CREATE STREAM sorders WITH (KAFKA_TOPIC='orders', VALUE_FORMAT='AVRO');

DESCRIBE EXTENDED SORDERS;
exit
EOF

# initialize db (may not be needed)
sqlite3 test.db '.tables'

# register connector NOTE: db path needs to be adjusted
curl -s -X POST -H  "Content-Type:application/json" \
    http://localhost:8083/connectors/ \
    -d '
{
    "name": "sorders_sqlite",
    "config": {
        "connector.class": "io.confluent.connect.jdbc.JdbcSinkConnector",
        "tasks.max": "1",
        "topics": "SORDERS",
        "connection.url": "jdbc:sqlite:/home/brandl/test.db",
        "auto.create": "true",
        "key.converter": "org.apache.kafka.connect.storage.StringConverter",
        "key.ignore": "true",
        "pk.mode": "none",
        "value.converter": "io.confluent.connect.avro.AvroConverter",
        "value.converter.schemas.enable": "false",
        "value.converter.schema.registry.url": "http://localhost:8081"
    }
}
' | jq .



# send example to orders topic to see if its propagated into stream
kafka-avro-console-producer \
 --broker-list localhost:9092 --topic orders \
 --property value.schema='{"type":"record","name":"myrecord","fields":[{"name":"id","type":"int"},{"name":"product", "type": "string"}, {"name":"quantity", "type": "int"}, {"name":"price",
 "type": "float"}]}' <<EOF
 {"id": 999, "product": "foo", "quantity": 99, "price": 501}
EOF

## check status of connector configuration
confluent local status sorders_sqlite

# peak into db
sqlite3 test.db '.tables'
```
The table is **not** created.

Interestingly, when doing the same again  but after setting a key on the stream (by repartitioning the stream), the jdbc sink is correctly populated with data:
```

# build another stream where a key is set
ksql <<EOF
CREATE STREAM SORDERS_PRODKEY
  WITH (TIMESTAMP='rowtime', VALUE_FORMAT='AVRO') AS
  SELECT *
  FROM sorders
  PARTITION BY product;

DESCRIBE EXTENDED SORDERS_PRODKEY;
exit
EOF

curl -s -X POST -H  "Content-Type:application/json" \
    http://localhost:8083/connectors/ \
    -d '
{
    "name": "sorders_prodkey_sqlite",
    "config": {
        "connector.class": "io.confluent.connect.jdbc.JdbcSinkConnector",
        "tasks.max": "1",
        "topics": "SORDERS_PRODKEY",
        "connection.url": "jdbc:sqlite:/home/brandl/test.db",
        "auto.create": "true",
        "key.converter": "org.apache.kafka.connect.storage.StringConverter",
        "key.ignore": "true",
        "pk.mode": "none",
        "value.converter": "io.confluent.connect.avro.AvroConverter",
        "value.converter.schemas.enable": "false",
        "value.converter.schema.registry.url": "http://localhost:8081"
    }
}
' | jq .



# send example to orders topic to see if its propagated into the db
kafka-avro-console-producer \
 --broker-list localhost:9092 --topic orders \
 --property value.schema='{"type":"record","name":"myrecord","fields":[{"name":"id","type":"int"},{"name":"product", "type": "string"}, {"name":"quantity", "type": "int"}, {"name":"price",
 "type": "float"}]}' <<EOF
 {"id": 999, "product": "foo", "quantity": 99, "price": 501}
EOF

confluent local status sorders_prodkey_sqlite

# peak into db
sqlite3 test.db '.tables'
sqlite3 test.db 'SELECT COUNT(*) from SORDERS_PRODKEY'
```
This time `SORDERS_PRODKEY` is correctly created in the sqlitedb and populated with data.

So summing up the issue, it seems mandatory to set a key when using a jdbc sink for a ksql stream, but it's not clear if this is intended (misconfiguration in the protocol from above) or a bug.

**Comment by holgerbrandl** [comment]
Thanks for your support. Feel welcome to move the issue as needed.

```
ksql> PRINT 'orders' FROM BEGINNING;
Format:AVRO
10/2/19, 8:37:21 PM UTC, null, {"id": 999, "product": "foo", "quantity": 100, "price": 50.0}
10/2/19, 8:37:33 PM UTC, null, {"id": 999, "product": "foo", "quantity": 100, "price": 50.0}
10/2/19, 8:38:26 PM UTC, null, {"id": 999, "product": "foo", "quantity": 100, "price": 50.0}
10/2/19, 8:38:52 PM UTC, null, {"id": 999, "product": "foo", "quantity": 100, "price": 50.0}
10/2/19, 8:39:22 PM UTC, null, {"id": 999, "product": "foo", "quantity": 100, "price": 50.0}
10/2/19, 8:39:31 PM UTC, null, {"id": 999, "product": "foo", "quantity": 100, "price": 50.0}
10/2/19, 8:39:40 PM UTC, null, {"id": 999, "product": "foo", "quantity": 100, "price": 50.0}
...
```

```
ksql> PRINT 'SORDERS_PRODKEY' FROM BEGINNING;
Format:AVRO
10/7/19, 10:14:19 AM UTC, foo, {"ID": 999, "PRODUCT": "foo", "QUANTITY": 99, "PRICE": 501.0}
10/7/19, 10:34:16 AM UTC, foo, {"ID": 999, "PRODUCT": "foo", "QUANTITY": 99, "PRICE": 501.0}
10/7/19, 10:34:27 AM UTC, foo, {"ID": 999, "PRODUCT": "foo", "QUANTITY": 99, "PRICE": 501.0}
10/7/19, 10:38:28 AM UTC, foo, {"ID": 999, "PRODUCT": "foo", "QUANTITY": 99, "PRICE": 501.0}
10/7/19, 10:40:13 AM UTC, foo, {"ID": 999, "PRODUCT": "foo", "QUANTITY": 99, "PRICE": 501.0}
10/7/19, 10:40:21 AM UTC, foo, {"ID": 999, "PRODUCT": "foo", "QUANTITY": 99, "PRICE": 501.0}
...
```
The `...` denote many more records with the same structure which I've omitted here.

And for completeness here are also the stream descriptions:

```
ksql> DESCRIBE EXTENDED SORDERS;

Name                 : SORDERS
Type                 : STREAM
Key field            :
Key format           : STRING
Timestamp field      : Not set - using <ROWTIME>
Value format         : AVRO
Kafka topic          : orders (partitions: 1, replication: 1)

 Field    | Type
--------------------------------------
 ROWTIME  | BIGINT           (system)
 ROWKEY   | VARCHAR(STRING)  (system)
 ID       | INTEGER
 PRODUCT  | VARCHAR(STRING)
 QUANTITY | INTEGER
 PRICE    | DOUBLE
--------------------------------------           
```


```
ksql> DESCRIBE EXTENDED SORDERS_PRODKEY;                                                   
                                                                                           
Name                 : SORDERS_PRODKEY                                                     
Type                 : STREAM                                                              
Key field            : PRODUCT                                                             
Key format           : STRING                                                              
Timestamp field      : ROWTIME                                                             
Value format         : AVRO                                                                
Kafka topic          : SORDERS_PRODKEY (partitions: 1, replication: 1)                     
                                                                                           
 Field    | Type                                                                           
--------------------------------------                                                     
 ROWTIME  | BIGINT           (system)                                                      
 ROWKEY   | VARCHAR(STRING)  (system)                                                      
 ID       | INTEGER                                                                        
 PRODUCT  | VARCHAR(STRING)                                                                
 QUANTITY | INTEGER                                                                        
 PRICE    | DOUBLE                                                                         
--------------------------------------                                                     
                                                                                           
Queries that write into this STREAM                                                        
-----------------------------------                                                        
CSAS_SORDERS_PRODKEY_7 : CREATE STREAM SORDERS_PRODKEY WITH (REPLICAS = 1, PARTITIONS = 1, 
= 'SORDERS_PRODKEY', TIMESTAMP = 'rowtime') AS SELECT *                                    
FROM SORDERS SORDERS                                                                       
PARTITION BY PRODUCT;                                                                      
                                                                                           
For query topology and execution plan please run: EXPLAIN <QueryId>                        
```

So for my newbie eyes, the key seems to be the only difference between the streams.

Feel welcome if you need further information to track down the source of the problem.

**Comment by agavra** [comment]
Ah that was pesky! I reproduced this locally and then realized that your in connector configuration you specified the source name, not the topic name 😨 

Your configuration should have `"topics": "orders",` instead of `"topics":"SORDERS"`:
```
{
    "name": "sorders_sqlite",
    "config": {
        "connector.class": "io.confluent.connect.jdbc.JdbcSinkConnector",
        "tasks.max": "1",
        "topics": "orders",
        "connection.url": "jdbc:sqlite:/home/brandl/test.db",
        "auto.create": "true",
        "key.converter": "org.apache.kafka.connect.storage.StringConverter",
        "key.ignore": "true",
        "pk.mode": "none",
        "value.converter": "io.confluent.connect.avro.AvroConverter",
        "value.converter.schemas.enable": "false",
        "value.converter.schema.registry.url": "http://localhost:8081"
    }
}
```

You may already know this, but it is important to note that `CREATE STREAM` and `CREATE TABLE` (without an `AS`) statements in KSQL don't "do" anything, they just import metadata as DDL.

**Comment by holgerbrandl** [comment]
Thanks a lot for your in-depth analysis of my issue and your solution. 

I was not aware of the difference in semantics of the two  `CREATE STREAM` flavours (from stream vs. from topic). Since both constructs seem to create a stream (from different types of inputs in my newbie understanding) it is rather unexpected that just the `CREATE STREAM AS SELECT ` is persisted into a topic and can be thus used as source for kafka-connect.

I think https://docs.confluent.io/current/ksql/docs/developer-guide/create-a-stream.html should clarify the difference better. Clearly, it mentions persistence but just to compare against non-persistent continuous queries (`SELECT ...`). The fact that a stream created from a topic is not a valid input for connect, but is just a metadata import could be emphasized better imho.

**Comment by agavra** [comment]
I agree this is confusing and it took me a while to wrap my head around the difference between topics and STREAM/TABLE. cc @JimGalasyn who might want to know about this

**Comment by JimGalasyn** [comment]
@agavra Thanks for the heads up! I opened KSQL-3323 to track.
